cache/src/main/java/io/envoyproxy/controlplane/cache/XdsRequest.java

```diff
@@ -0,0 +1,88 @@
+package io.envoyproxy.controlplane.cache;
+
+import com.google.auto.value.AutoValue;
+import com.google.protobuf.ProtocolStringList;
+import io.envoyproxy.envoy.api.v2.DiscoveryRequest;
+import javax.annotation.Nullable;
+
+/**
+ * XdsRequest wraps a v2 or v3 DiscoveryRequest of and provides common methods as a
+ * workaround to the proto messages not implementing a common interface that can be used to
+ * abstract away xDS version. XdsRequest is passed around the codebase through common code,
+ * however the callers that need the raw request from it have knowledge of whether it is a v2 or
+ * a v3 request.
+ */
+@AutoValue
+public abstract class XdsRequest {
```

Comment:
This class is basically a oneof of v2 request or v3 request, and delegates methods to the right one.